keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,1089 @@
+// stores the corresponding attributes. The spec is read-only.
+type CSIStoragePool struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use sp-<uuid>.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	Spec   CSIStoragePoolSpec
+	Status CSIStoragePoolStatus
+}
+
+// CSIStoragePoolSpec contains the constant attributes of a CSIStoragePool.
+type CSIStoragePoolSpec struct {
+	// The CSI driver that provides access to the storage pool.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+}
+
+// CSIStoragePoolStatus contains runtime information about a CSIStoragePool.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster as identified by NodeTopology. If not set, the pool is assumed
+// to be available in the entire cluster.
+//
+// It is expected to be extended with other
+// attributes which do not depend on the storage class, like health of
+// the pool. Therefore it has the list of
+// `CSIStorageByClass` instances instead of just the capacity
+// and the storage class being in the spec.
+type CSIStoragePoolStatus struct {
+	// NodeTopology can be used to describe a storage pool that is available
+	// only for nodes matching certain criteria.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// Some information, like the actual usable capacity, may
+	// depend on the storage class used for volumes.
+	//
+	// +patchMergeKey=storageClassName
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=storageClassName
+	// +optional
+	Classes []CSIStorageByClass `patchStrategy:"merge" patchMergeKey:"storageClassName" json:"classes,omitempty" protobuf:"bytes,4,opt,name=classes"`
+}
+
+// CSIStorageByClass contains information that applies to one storage
+// pool of a CSI driver when using a certain storage class.
+type CSIStorageByClass struct {
+	// The storage class name matches the name of some actual
+	// `StorageClass`, in which case the information applies when
+	// using that storage class for a volume. There is also one
+	// special name:
+	// - <ephemeral> for storage used by ephemeral inline volumes (which
+	//   don't use a storage class)
+	StorageClassName string `json:"storageClassName" protobuf:"bytes,1,name=storageClassName"`
+
+	// Capacity is the value reported by the CSI driver in its GetCapacityResponse.
+    // Depending on how the driver is implemented, this might be the total
+    // size of the available storage which is only available when allocating
+    // multiple smaller volumes ("total available capacity") or the
+    // actual size that a volume may have ("maximum volume size").
```

Comment:
This pretty much guarantees this feature will can not be used together with Cluster Autoscaler. Autoscaler works by looking at all pending pods and simulating the result of scheduling them if more nodes were added.
With the above semantics you can't simulate the result of scheduling multiple pods, as you can't predict how scheduling one pod will change Capacity and how that will impact your ability to schedule other pods on the same node. 

We already have a similar issue with local PVs not working with autoscaler. In this case we have a general idea how to fix it (though it will require significant changes across autoscaler codebase, so it's going to take a while before it happens). 
It's not relevant for on-prem users, but CA is very widely used on all major clouds. With local PVs we're seeing a lot of issues where cloud users can't adopt them because of incompatibility with autoscaler and i expect the same will happen with this feature.

We should either try to update the design to make it predictable how scheduling one pod changes capacity or explicitly call Cluster Autoscaler support a non-goal and accept the impact it will have on usability of this feature in cloud environments.

Reply:
> This is a rather strong statement.

I think it's justified. Autoscaler works by predicting what will be the result of scheduling multiple pods (often thousands of them) on nodes that don't even exist yet. Note that it doesn't re-implement scheduling logic - it just feeds in-memory node and pod objects to scheduler Filters and checks if it gets a yes or no answer. Than it applies any side effect needed to simulate 'scheduling' one pod and sees how that impacts schedulability of other pods. 
Except in this case there isn't a well defined semantic for applying side effects of scheduling one pod. If I understand correctly it's a runtime decision that cannot be predicted ahead of time. If that is the case it just can't possibly work with autoscaler.

> Modeling in Kubernetes for arbitrary CSI drivers how volume creation and deletion affects capacity is IMHO futile.

I see your point there. But there is a trade off between customizability and automation. Maybe we can consider some sort of middle ground where at least simple cases can still be supported by autoscaler. I'm not sure I know enough about CSI to suggest a specific solution, but maybe we can add optional TotalCapacity for cases where we can simply subtract from it or something along those lines?

> For the first iteration (i.e. alpha in 1.19) let's exclude Autoscaler and then decide whether it needs to remain out of scope permanently or what can be done for it.

That is the answer I get every time I bring autoscaler in any feature discussion :) 
And now I have plenty of examples for how it doesn't work. Scheduler framework allows for a lot, but autoscaler adds limitations on how you can use scheduler (ex. Filters can't have internal state or side effects, unless there is a way for autoscaler to control both). This is not typically something that can be easily added later on.

Reply:
> I removed it in e02fce0#diff-f3298a4cee270f1b16246ce3b5504ab1 to address this feedback: #1353 (comment)

I would say we have a justification for having two fields now: simulating scheduling multiple pods in a new node. I would say that TotalCapacity should be mandatory, and "max volume size" optional (if it's not present, it's assumed to be total capacity).

> Modeling in Kubernetes for arbitrary CSI drivers how volume creation and deletion affects capacity is IMHO futile.

Let's discuss a single pod for now. If the capacity published by the CSI driver is correct, we should be able to simulate the creation of a new node for that pod.
Now, if we introduce multiple pods, it is possible that a new node will be able to accommodate all of them as autoscaler is predicting. If it does, the job is done. If it doesn't, autoscaler would eventually trigger the creation of more nodes.
I know this is not ideal, but then it's up to the CSI driver to behave as closely as it can to the capacity it publishes so that autoscaler works perfectly with it.

Reply:
> I would say we have a justification for having two fields now

My understanding was that having TotalCapacity would mean that you can subtract from it when scheduling pods? If it does than it would indeed solve the issue.

>  If the capacity published by the CSI driver is correct, we should be able to simulate the creation of a new node for that pod.
Now, if we introduce multiple pods, it is possible that a new node will be able to accommodate all of them as autoscaler is predicting.

Yeah, going one-by-one would work on some level. But if autoscaler think all those pods could fit on one node it is likely to create huge node with enough CPU for all of them, only to find out only a single pod can in fact run on this node. It is _some_ level of support, but not a very good one. However, if we know that we can predict behavior in some cases (ex. because total behaves like cpu or memory in some case), we could have much better support for those cases.

Reply:
That's good to hear. Moving forward (beyond the scope of this KEP), there could be fields for "min volume size" and/or "volume chunk size" (meaning that sizes can only be multiples of it). I think that should cover most simulations.

Reply:
@msau42: you originally proposed having two fields and extending the CSI spec so that we actually can retrieve both, then later suggested to just stick with a single field for now because we don't have that extended CSI yet.

Do you still prefer to do an initial alpha with just one field or shall we switch back to two fields and later figure out how to populate them?

I'm leaning towards the latter, but I I'm also fine with just one field for now.

Reply:
I discussed this a bit with @alculquicondor and @saad-ali. 

From the scheduling side, there are scalability/performance concerns when you have high pod churn requesting these types of volumes (and yes, when you add in the ephemeral scratch space use case, it's realistic to have a high churn of pods) because the scheduler will have to retry many times before capacity information is updated. Updating capacity information frequently through the API is not ideal because it puts a lot of pressure on the API server.

From the autoscaling side, support for this feature will be challenging because it cannot use the information as defined in the API, and would have to potentially use other heuristics and write custom scheduler logic potentially for specific drivers to work with it.

From the storage side, modeling total capacity is hard and only going to work under certain scenarios: the storage system is contiguous, storageclasses do not share the same physical capacity, etc. So adding total capacity would only be applicable to a subset of storage systems. We have already debated this in the CSI spec: https://github.com/container-storage-interface/spec/issues/301. So if we add a total capacity field, we would need to carefully think through all the constraints and possibilities of what could go wrong if one of these constraints wasn't met. In addition, if we add the field to Kubernetes but not CSI, then that requires drivers to be non-compliant if they want to support this.

So I think there's still some more investigation we need to do on this. My proposed next steps are:
* Document all these downsides in the KEP and make investigation of these issues a beta criteria
* Write a CSI driver that we believe does not have the problems of modeling total capacity (IMO, I think a simple LVM local driver could fit this use case)
* Run this driver through stress scenarios and measure the impact on scheduling latency
* Prototype autoscaling support with some simple heuristic and see where it fails

We can use the results of the investigation to help us determine if these downsides are indeed a problem. We can also potentially use the driver to prototype a solution using total capacity and rerun the tests to see if it actually solve our problems. I think having real data and a prototype will allow us to make a more informed argument for the next direction.

Of course, we're going to need support from sig storage, scheduling, and autoscaling to accomplish all this.

@pohly @alculquicondor @MaciekPytel wdyt of this plan?

Reply:
> From the scheduling side, there are scalability/performance concerns when you have high pod churn requesting these types of volumes (and yes, when you add in the ephemeral scratch space use case, it's realistic to have a high churn of pods)

This is an assumption that we'll have to check by querying users. Provisioning volumes and more importantly, *using* large volumes is going to be slow, which implies that pods are likely to be long-lived and thus only need to be scheduled infrequently. For pods which need small amounts of scratch space, the existing `EmptyDir` volume source is probably better. For example, the main use case I had in mind for generic inline volumes is something like memcached, which is not something that is going to be started up and shut down frequently.

> Document all these downsides in the KEP and make investigation of these issues a beta criteria

I'll try to capture more of the downsides. If I miss one, please remind me...

> Write a CSI driver that we believe does not have the problems of modeling total capacity (IMO, I think a simple LVM local driver could fit this use case)

Let me offer PMEM-CSI here. It has one mode where it uses LVM and with some small tweak it will be possible to use it with non-PMEM storage. Using it has the advantage that I can make changes both in the driver and in Kubernetes because I work on both projects.

> Run this driver through stress scenarios and measure the impact on scheduling latency

Absolutely. It has been my mantra all along that we need to build something and then evaluate how well it works. These advance discussions are useful, but only up to a certain point. Sooner or later we end up speculating.

>  Prototype autoscaling support with some simple heuristic and see where it fails

Yes. I think this can be made to work, we just need to explore.

> wdyt of this plan?

Makes sense to me as long as we agree to do this work *in* Kubernetes and start *now*, i.e. target an initial alpha implementation in 1.19. I don't want to maintain a fork of Kubernetes with this feature (because I fear that it'll just not get the necessary attention and because of the additional work that comes with keeping a fork up-to-date) and I don't want to delay (because that implies forking and because the sooner we start some real work, the sooner we'll be able to finish).

Reply:
> I'll try to capture more of the downsides. If I miss one, please remind me...

Done, see b75421d.

I also renamed the directory because it was pointed out to me on Slack that just the issue number isn't what the KEP template asks for.